docs/usage/configuration-options.md

```diff
@@ -2235,22 +2235,64 @@ e.g.
 
 The `postUpgradeTasks` configuration consists of three fields:
 
-### commands
+### commands (postUpgradeTasks)
 
 A list of commands that are executed after Renovate has updated a dependency but before the commit is made.
 
-You can use variable templating in your commands if [`allowPostUpgradeCommandTemplating`](https://docs.renovatebot.com/self-hosted-configuration/#allowpostupgradecommandtemplating) is enabled.
+You can use variable templating in your commands if [`allowUpgradeCommandTemplating`](https://docs.renovatebot.com/self-hosted-configuration/#allowUpgradeCommandTemplating) is enabled.
 
-### fileFilters
+### fileFilters (postUpgradeTasks)
 
 A list of glob-style matchers that determine which files will be included in the final commit made by Renovate.
 
-### executionMode
+### executionMode (postUpgradeTasks)
 
 Defaults to `update`, but can also be set to `branch`.
 This sets the level the postUpgradeTask runs on, if set to `update` the postUpgradeTask will be executed for every dependency on the branch.
 If set to `branch` the postUpgradeTask is executed for the whole branch.
 
+## preUpgradeTasks
+
+<!-- prettier-ignore -->
+!!! note
+    Pre-upgrade tasks can only be used on self-hosted Renovate instances.
+
+Pre-upgrade tasks are commands that are executed by Renovate before a dependency has been updated but after the checkout on the dependency branch.
+The intention is to run any additional command line tools that would modify existing files or generate new files when a dependency changes.
+
+Each command must match at least one of the patterns defined in `allowedUpgradeCommands` (a global-only configuration option) in order to be executed.
+If the list of allowed tasks is empty then no tasks will be executed.
+
+e.g.
+
+```json
+{
+  "preUpgradeTasks": {
+    "commands": ["xcodegen generate"],
+    "fileFilters": [],
+    "executionMode": "update"
+  }
+}
+```
```

Comment:
It's not clear to me if this is an example of the _good_ way, or the _bad_ way to configure the bot.